cirq-core/cirq/ops/kraus_channel.py

```diff
@@ -0,0 +1,105 @@
+from typing import Any, Dict, Iterable, Tuple, Union
+import numpy as np
+
+from cirq import linalg, protocols, value
+from cirq._compat import proper_repr
+from cirq.ops import raw_types
+
+
+# TODO(#3241): support qudits and non-square operators.
+class KrausChannel(raw_types.Gate):
+    """A generic channel that can record the index of its selected operator.
+
+    Args:
+        kraus_ops: a list of Kraus operators, formatted as numpy array.
+            Currently, only square-matrix operators on qubits (not qudits) are
+            supported by this type.
+        key: an optional measurement key string for this channel. Simulations
+            which select a single Kraus operator to apply will store the index
+            of that operator in the measurement result list with this key.
+        validate: if True, validate that `kraus_ops` describe a valid channel.
+            This validation can be slow; prefer pre-validating if possible.
+    """
+
+    def __init__(
+        self,
+        kraus_ops: Iterable[np.ndarray],
+        key: Union[str, value.MeasurementKey, None] = None,
+        validate: bool = False,
+    ):
+        kraus_ops = list(kraus_ops)
+        if not kraus_ops:
+            raise ValueError('KrausChannel must have at least one operation.')
+        num_qubits = np.log2(kraus_ops[0].size) / 2
+        if not num_qubits.is_integer():
+            raise ValueError(
+                f'Input Kraus ops of shape {kraus_ops[0].shape} does not '
+                'represent an operator over qubits.'
+            )
+        self._num_qubits = int(num_qubits)
+        for i, op in enumerate(kraus_ops):
+            if not op.size == kraus_ops[0].size:
+                raise ValueError(
+                    'Inconsistent Kraus operator sizes: '
+                    f'op[0]: {kraus_ops[0].size}, op[{i}]: {op.size}'
+                )
+        if validate and not linalg.is_cptp(kraus_ops=kraus_ops):
+            raise ValueError('Kraus operators do not describe a CPTP map.')
+        self._kraus_ops = kraus_ops
+        if not isinstance(key, value.MeasurementKey) and key is not None:
+            key = value.MeasurementKey(key)
+        self._key = key
+
+    @staticmethod
+    def from_channel(
+        channel: 'protocols.SupportsChannel', key: Union[str, value.MeasurementKey, None] = None
+    ):
+        """Creates a copy of a channel with the given measurement key."""
+        return KrausChannel(kraus_ops=list(protocols.kraus(channel)), key=key)
+
+    def __eq__(self, other) -> bool:
+        if not isinstance(other, KrausChannel):
+            return NotImplemented
+        if self._key != other._key:
+            return False
+        return np.allclose(self._kraus_ops, other._kraus_ops)
```

Comment:
Kraus representation is not unique, so when measurement key is unspecified this check is insufficient. Instead, we should compare, e.g. the Choi representations which determine the channel uniquely. OTOH, if the measurement key is specified, then different sets of Kraus operators that give rise to the same Choi will generally describe different channels and then the current logic almost applies. I say "almost" since there is the caveat that one could apply any complex phase factor to the Kraus operators and get equivalent `KrausChannel`.

I think we should do something like this:

```python
if self._key is None and other._key is None:
    return np.allclose(cirq.kraus_to_choi(self._kraus_ops), cirq.kraus_to_choi(other._kraus_ops))
return all(cirq.equal_up_to_global_phase(k1, k2) for k1, k2 in zip(self._kraus_ops, other._kraus_ops))
```

This assumes that if measurement key is specified then `KrausChannel` instances that differ in the order of otherwise equivalent Kraus operators are to be considered unequal. This makes sense since they generate different measurement outcomes.

It'd be good to add unit tests for both cases above. You can use for example these two sets of Kraus operators for the phase damping channel:

1. I/2 and Z/2.
2. [[1, 0], [0, 0]] and [[0, 0], [0, 1]].

Reply:
Yes, defining a separate method is definitely fine (and TODO is also fine), but please add a comment to tell users that `KrausChannel` instances that describe the same linear map may be unequal if they're specified via different Kraus operators.

If we go this route then we should probably leave this as is and not use `equal_up_to_global_phase`. Otherwise it feels like a half-baked attempt at "behavioral" equality (global phase freedom is just a special case of unitary freedom in Kraus representation which is given by diagonal unitaries).

Reply:
Added a TODO for this and left the `np.allclose` in `__eq__`.